### PR TITLE
Refactor Rag notebook

### PR DESCRIPTION
**Pull request by richardsliu**
Refactored the RAG notebook to be more modular and documented. 

* Moved the CloudSQL code into the notebook instead of running on a Ray Clusters
* Execute remote Ray tasks instead of using a job submit with busy waiting
* Modified the CloudSQL code to use bulk insert

Fixes https://github.com/GoogleCloudPlatform/ai-on-gke/issues/425

Preview: https://github.com/GoogleCloudPlatform/ai-on-gke/blob/rag-notebook/applications/rag/example_notebooks/rag-kaggle-ray-sql-refactored.ipynb

### CODE REVIEW

File: modules/jupyter/jupyter_config/config-selfauth-autopilot.yaml [hunk 1]
```diff
@@ -88,6 +88,7 @@ singleuser:
   extraEnv:
     # Used for GCSFuse to set the ephemeral storage as the home directory. If not set, it will show a permission error on the pod log when using GCSFuse.
     JUPYTER_ALLOW_INSECURE_WRITES: "true"
+    CLOUDSQL_INSTANCE_CONNECTION_NAME: ${cloudsql_instance_connection_name}
```
Comment (andrewsykim) [inline modules/jupyter/jupyter_config/config-selfauth-autopilot.yaml:91]: btw. this is going to add the CLOUDSQL env var and secrets for the Jupyter-only installation (not RAG)

Comment (andrewsykim) [reply modules/jupyter/jupyter_config/config-selfauth-autopilot.yaml:91]: (likewise for the secret mounts)

File: applications/rag/example_notebooks/rag-kaggle-ray-sql-refactored.ipynb [hunk 2]
```diff
@@ -0,0 +1,449 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "a814e91b-3afe-4c28-a3d6-fe087c7af552",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "!pip install ray[default]==2.9.3 kaggle==1.6.6\n",
+    "!pip install langchain==0.1.10 ray==2.9.3 datasets sentence-transformers\n",
```
Comment (andrewsykim) [inline applications/rag/example_notebooks/rag-kaggle-ray-sql-refactored.ipynb:51]: I think there were some issues in the past with some of these packages taking way too long to install. Specifically sentence-transformers. Are we going to bake the dependencies into the jupyter image?

File: applications/rag/example_notebooks/rag-kaggle-ray-sql-refactored.ipynb [hunk 3]
```diff
@@ -0,0 +1,449 @@
+{
```
Comment (andrewsykim) [inline applications/rag/example_notebooks/rag-kaggle-ray-sql-refactored.ipynb:1]: Can you name this file `rag-kaggle-ray-sql-interactive.ipynb` instead?

File: modules/jupyter/jupyter_config/config-selfauth-autopilot.yaml (hunk 1 above)
Comment (andrewsykim) [reply modules/jupyter/jupyter_config/config-selfauth-autopilot.yaml:91]: ^ we spoke offline -- I meant that jupyer only installs will set these environment varibles

**Review by andrewsykim — COMMENTED** [review]
Overall LGTM, this is a great improvement!

I think the following issues need to be addressed if we're going to cherry-pick this into release-1.1:
* Dependency install time and whether a custom image is warranted. 
* Leaking CloudSQL env vars and secret mounts into all Jupyter installments -- these should only be set when using Jupyter for RAG.
* If we choose to keep both versions of the notebook, please add the markdown descriptions in the other notebook as well

File: modules/jupyter/jupyter_config/config-selfauth-autopilot.yaml (hunk 1 above)
Comment (richardsliu) [reply modules/jupyter/jupyter_config/config-selfauth-autopilot.yaml:91]: Made these variables default to empty string. If not explicitly passed, the env var will be empty.

File: applications/rag/example_notebooks/rag-kaggle-ray-sql-refactored.ipynb (hunk 2 above)
Comment (richardsliu) [reply applications/rag/example_notebooks/rag-kaggle-ray-sql-refactored.ipynb:51]: I've tested and it turns out the notebook does need to install langchain and sentence-transformers. I can see if we could use a custom image here to skip these pip installs.

**Comment by richardsliu** [comment]
> * Dependency install time and whether a custom image is warranted.

The notebook needs to install `sentence-transformer` and `langchain`. I'll see if we can use a custom image to reduce the install time.

> * Leaking CloudSQL env vars and secret mounts into all Jupyter installments -- these should only be set when using Jupyter for RAG (likewise for Ray ref: [CLOUDSQL_INSTANCE_CONNECTION_NAME env in RayCluster should only be configured for RAG deployments   #435](https://github.com/GoogleCloudPlatform/ai-on-gke/issues/435)).

This is now fixed.

> * If we choose to keep both versions of the notebook, please add the markdown descriptions in the other notebook as well

I prefer to keep this PR limited to changes relevant for the new notebook. Adding markdown to the older notebook wouldn't be too interesting since all the code is executed in one cell.

File: applications/rag/example_notebooks/rag-kaggle-ray-sql-refactored.ipynb (hunk 2 above)
Comment (richardsliu) [reply applications/rag/example_notebooks/rag-kaggle-ray-sql-refactored.ipynb]: Changed to a custom image with dependencies baked in. Removed this section from the notebook.

File: applications/rag/example_notebooks/rag-kaggle-ray-sql-refactored.ipynb [hunk 4]
```diff
@@ -0,0 +1,449 @@
...
+   "id": "ba9551ec-883e-4bde-9f12-f663aedc12e5",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import ray\n",
+    "\n",
+    "ray.init(\n",
+    "    address=\"ray://ray-cluster-kuberay-head-svc:10001\",\n",
+    "    runtime_env={\n",
+    "        \"pip\": [               \n",
```
[the remark this answers is not in the document]
Comment (richardsliu) [reply applications/rag/example_notebooks/rag-kaggle-ray-sql-refactored.ipynb]: I think Ray will skip pip installing these if the image is already on the node. When I tested this, this step usually finishes in a few seconds.

File: applications/rag/example_notebooks/rag-kaggle-ray-sql-interactive.ipynb [hunk 5]
```diff
@@ -0,0 +1,425 @@
...
+   "metadata": {},
+   "source": [
+    "# RAG-on-GKE Application\n",
+    "\n",
+    "This is a Python notebook for generating the vector embeddings used by the RAG on GKE application. For full information, please checkout the GitHub documentation [here](https://github.com/GoogleCloudPlatform/ai-on-gke/blob/main/applications/rag/README.md).\n",
+    "\n",
+    "\n",
+    "## Setup Kaggle Credentials\n",
+    "\n",
+    "First we will setup your Kaggle credentials. Replace the following with your own settings from the Kaggle web page. Navigate to https://www.kaggle.com/settings/account and generate an API token to be used to setup the env variable. See https://www.kaggle.com/docs/api#authentication how to create one."
```
Comment (imreddy13) [inline applications/rag/example_notebooks/rag-kaggle-ray-sql-interactive.ipynb:15]: nit: "to be used to setup the env variable" is kind of weird. Maybe just have the instructions for how to create Kaggle creds and ask users to replace the specific variables (KAGGLE_USERNAME and KAGGLE_KEY) in the cell below with their Kaggle creds?

File: modules/jupyter/jupyter_config/config-selfauth-autopilot.yaml [hunk 6]
```diff
@@ -111,9 +112,17 @@ singleuser:
...
             mountOptions: "implicit-dirs,uid=1000,gid=100"
+      - name: secret-volume
+        secret:
+          secretName: ${secret_name}
+          optional: true
+
     extraVolumeMounts:
         - name: data-vol
-          mountPath: /persist-data
+          mountPath: /data
```
Comment (imreddy13) [inline modules/jupyter/jupyter_config/config-selfauth-autopilot.yaml:122]: we don't need persist-data anymore? awesome! this always confused me.

File: modules/jupyter/main.tf [hunk 7]
```diff
@@ -17,6 +17,7 @@ data "google_project" "project" {
 }
 
 locals {
+  cloudsql_instance_connection_name = var.cloudsql_instance_name != "" ? format("%s:%s:%s", var.project_id, var.db_region, var.cloudsql_instance_name) : ""
```
Comment (imreddy13) [inline modules/jupyter/main.tf:20]: could you please try to run tf apply from /applications/jupyter to ensure that the secrets mounting isn't breaking just the jupyter installation? it should be fairly quick to test

File: modules/jupyter/variables.tf [hunk 8]
```diff
@@ -17,6 +17,18 @@ variable "namespace" {
   description = "Kubernetes namespace where resources are deployed"
 }
 
+variable "notebook_image" {
+  type        = string
+  description = "Jupyter notebook image name"
+  default     = "jupyter/tensorflow-notebook"
```
Comment (imreddy13) [inline modules/jupyter/variables.tf:23]: In addition to defaulting, can you also explicitly set this value from /applications/jupyter/main.tf so it's clear which app is using which image?

File: modules/jupyter/variables.tf [hunk 9]
```diff
@@ -17,6 +17,18 @@ variable "namespace" {
...
+}
+
+variable "notebook_image_tag" {
+  type        = string
+  description = "Jupyter notebook image tag"
+  default     = "python-3.10"
```
Comment (imreddy13) [inline modules/jupyter/variables.tf:29]: same

File: applications/rag/example_notebooks/rag-kaggle-ray-sql-interactive.ipynb (hunk 5 above)
Comment (richardsliu) [reply applications/rag/example_notebooks/rag-kaggle-ray-sql-interactive.ipynb]: Merged the cells.

File: modules/jupyter/jupyter_config/config-selfauth-autopilot.yaml (hunk 6 above)
Comment (richardsliu) [reply modules/jupyter/jupyter_config/config-selfauth-autopilot.yaml:122]: Glad you agree.

File: modules/jupyter/variables.tf (hunk 8 above)
Comment (richardsliu) [reply modules/jupyter/variables.tf:23]: Done.

File: modules/jupyter/main.tf (hunk 7 above)
Comment (richardsliu) [reply modules/jupyter/main.tf:20]: Good call. Fixed an error.

**Comment by andrewsykim** [comment]
> I prefer to keep this PR limited to changes relevant for the new notebook. Adding markdown to the older notebook wouldn't be too interesting since all the code is executed in one cell.

Should we simplify and remove the old notebook or keep both? With the changes in this PR, do both notebooks work?

**Comment by richardsliu** [comment]
> > I prefer to keep this PR limited to changes relevant for the new notebook. Adding markdown to the older notebook wouldn't be too interesting since all the code is executed in one cell.
> 
> Should we simplify and remove the old notebook or keep both? With the changes in this PR, do both notebooks work?

I fixed some minor issues in the old notebook and added a header. We should probably keep it as a backup.

**Review by imreddy13 — APPROVED** [review]
Thank you!

**Pull request merged**